File: lua/filefs/format.lua
-- FileFS on-disk format helpers (512-byte blocks, magic 78 11 45 14).

local Format = {}

Format.BLOCK_SIZE = 512
Format.BLOCK_ITEM_MAX_COUNT = 20
Format.BLOCK_HEAD = 12
Format.BLOCK_NAME_MAX_SIZE = 14
Format.ENTRY_SIZE = 25
Format.ROOT_BLOCKINDEX = 1
Format.DATA_PER_BLOCK = Format.BLOCK_SIZE - Format.BLOCK_HEAD
Format.MAGIC = { 0x78, 0x11, 0x45, 0x14 }

local function empty_block()
  local b = {}
  for i = 1, Format.BLOCK_SIZE do
    b[i] = 0
  end
  return b
end

local function write_u32_le(bytes, offset0, value)
  -- offset0 is 0-based
  local i = offset0 + 1
  local v = value & 0xFFFFFFFF
  bytes[i] = v & 0xFF
  bytes[i + 1] = (v >> 8) & 0xFF
  bytes[i + 2] = (v >> 16) & 0xFF
  bytes[i + 3] = (v >> 24) & 0xFF
end

local function read_u32_le(bytes, offset0)
  local i = offset0 + 1
  return bytes[i]
    | (bytes[i + 1] << 8)
    | (bytes[i + 2] << 16)
    | (bytes[i + 3] << 24)
end

local function write_u16_le(bytes, offset0, value)
  local i = offset0 + 1
  local v = value & 0xFFFF
  bytes[i] = v & 0xFF
  bytes[i + 1] = (v >> 8) & 0xFF
end

local function read_u16_le(bytes, offset0)
  local i = offset0 + 1
  return bytes[i] | (bytes[i + 1] << 8)
end

local function write_fixed_name(bytes, offset0, name)
  assert(#name <= Format.BLOCK_NAME_MAX_SIZE, "name exceeds 14 bytes: " .. name)
  for i = 0, Format.BLOCK_NAME_MAX_SIZE - 1 do
    bytes[offset0 + 1 + i] = 0
  end
  for i = 1, #name do
    bytes[offset0 + i] = string.byte(name, i)
  end
end

local function read_fixed_name(bytes, offset0)
  local chars = {}
  for i = 0, Format.BLOCK_NAME_MAX_SIZE - 1 do
    local b = bytes[offset0 + 1 + i]
    if b == 0 then
      break
    end
    chars[#chars + 1] = string.char(b)
  end
  return table.concat(chars)
end

local function read_entry(bytes, offset0)
  return {
    is_file = (bytes[offset0 + 1] & 0x01) == 1,
    name = read_fixed_name(bytes, offset0 + 1),
    start = read_u32_le(bytes, offset0 + 1 + Format.BLOCK_NAME_MAX_SIZE),
    stop = read_u32_le(bytes, offset0 + 1 + Format.BLOCK_NAME_MAX_SIZE + 4),
    offset = read_u16_le(bytes, offset0 + 1 + Format.BLOCK_NAME_MAX_SIZE + 8),
  }
end

local function ceil_div(a, b)
  return math.floor((a + b - 1) / b)
end

function Format.new_dir(name, parent)
  return {
    kind = "dir",
    name = name,
    parent = parent,
    children = {}, -- ordered array of {name=, node=} plus map by name
    child_map = {},
  }
end

function Format.new_file(name, parent, data)
  return {
    kind = "file",
    name = name,
    parent = parent,
    data = data or {},
  }
end

function Format.dir_set_child(dir, name, node)
  if dir.child_map[name] then
    for i, entry in ipairs(dir.children) do
      if entry.name == name then
        dir.children[i] = { name = name, node = node }
        break
      end
    end
  else
    dir.children[#dir.children + 1] = { name = name, node = node }
  end
  dir.child_map[name] = node
  node.name = name
  node.parent = dir
end

function Format.dir_remove_child(dir, name)
  local node = dir.child_map[name]
  if not node then
    return nil
  end
  dir.child_map[name] = nil
  for i, entry in ipairs(dir.children) do
    if entry.name == name then
      table.remove(dir.children, i)
      break
    end
  end
  return node
end

function Format.dir_get_child(dir, name)
  return dir.child_map[name]
end

local function copy_bytes(src)
  local dst = {}
  for i = 1, #src do
    dst[i] = src[i]
  end
  return dst
end

function Format.deep_copy_tree(root, cwd)
  local map = {}

  local function copy_dir(source, parent)
    local copy = Format.new_dir(source.name, parent)
    map[source] = copy
    for _, entry in ipairs(source.children) do
      local child = entry.node
      if child.kind == "dir" then
        Format.dir_set_child(copy, entry.name, copy_dir(child, copy))
      else
        Format.dir_set_child(copy, entry.name, Format.new_file(entry.name, copy, copy_bytes(child.data)))
      end
    end
    return copy
  end

  local root_copy = copy_dir(root, nil)
  return {
    root = root_copy,
    cwd = map[cwd] or root_copy,
  }
end

function Format.build_absolute_path(dir)
  if dir.parent == nil then
    return "/"
  end
  local parts = {}
  local current = dir
  while current and current.parent do
    table.insert(parts, 1, current.name)
    current = current.parent
  end
  return "/" .. table.concat(parts, "/") .. "/"
end

function Format.list_directory_entries(dir)
  local FileType = require("filefs.types").FileType
  local entries = {}
  local dot_type = dir.parent == nil and FileType.ROOT or FileType.DIR
  entries[#entries + 1] = { type = dot_type, name = "." }
  entries[#entries + 1] = { type = dot_type, name = ".." }
  for _, entry in ipairs(dir.children) do
    local typ = entry.node.kind == "dir" and FileType.DIR or FileType.FILE
    entries[#entries + 1] = { type = typ, name = entry.name }
  end
  return entries
end

function Format.serialize_tree(root)
  local addresses = {} -- node -> address

  local function assign(node, next_index)
    if node.kind == "dir" then
      local entry_count = 2 + #node.children
      local block_count = ceil_div(entry_count, Format.BLOCK_ITEM_MAX_COUNT)
      local start = next_index
      local stop = start + block_count - 1
      local remainder = entry_count % Format.BLOCK_ITEM_MAX_COUNT
      local offset = remainder == 0 and Format.BLOCK_SIZE
        or (Format.BLOCK_HEAD + remainder * Format.ENTRY_SIZE)
      addresses[node] = {
        start = start,
        stop = stop,
        offset = offset,
        block_count = block_count,
      }
      local cursor = stop + 1
      for _, entry in ipairs(node.children) do
        cursor = assign(entry.node, cursor)
      end
      return cursor
    end

    if #node.data == 0 then
      addresses[node] = { start = 0, stop = 0, offset = 0, block_count = 0 }
      return next_index
    end

    local block_count = ceil_div(#node.data, Format.DATA_PER_BLOCK)
    local start = next_index
    local stop = start + block_count - 1
    local remainder = #node.data % Format.DATA_PER_BLOCK
    local offset = remainder == 0 and Format.BLOCK_SIZE
      or (Format.BLOCK_HEAD + remainder)
    addresses[node] = {
      start = start,
      stop = stop,
      offset = offset,
      block_count = block_count,
    }
    return stop + 1
  end

  local total_blocks = assign(root, Format.ROOT_BLOCKINDEX)
  local blocks = {}
  for i = 0, total_blocks - 1 do
    blocks[i] = empty_block()
  end

  blocks[0][1] = Format.MAGIC[1]
  blocks[0][2] = Format.MAGIC[2]
  blocks[0][3] = Format.MAGIC[3]
  blocks[0][4] = Format.MAGIC[4]
  write_u32_le(blocks[0], 4, total_blocks)
  write_u32_le(blocks[0], 8, 0)

  local function serialize_file(file)
    local address = addresses[file]
    if address.start == 0 then
      return
    end
    local source_offset = 0
    for block_index = address.start, address.stop do
      local block = blocks[block_index]
      local next = block_index < address.stop and (block_index + 1) or 0
      local prev = block_index > address.start and (block_index - 1) or 0
      write_u32_le(block, 4, next)
      write_u32_le(block, 8, prev)
      local end_exclusive = math.min(source_offset + Format.DATA_PER_BLOCK, #file.data)
      for i = source_offset + 1, end_exclusive do
        block[Format.BLOCK_HEAD + (i - source_offset)] = file.data[i]
      end
      source_offset = end_exclusive
    end
  end

  local function serialize_dir(dir)
    local address = addresses[dir]
    local entries = {}
    entries[#entries + 1] = {
      is_file = false,
      name = ".",
      start = address.start,
      stop = address.stop,
      offset = address.offset,
    }
    local parent_start = 0
    if dir.parent then
      parent_start = addresses[dir.parent].start
    end
    entries[#entries + 1] = {
      is_file = false,
      name = "..",
      start = parent_start,
      stop = 0,
      offset = 0,
    }
    for _, child_entry in ipairs(dir.children) do
      local child = child_entry.node
      local child_address = addresses[child]
      entries[#entries + 1] = {
        is_file = child.kind == "file",
        name = child.name,
        start = child_address.start,
        stop = child_address.stop,
        offset = child_address.offset,
      }
    end

    for block_offset = 0, address.block_count - 1 do
      local block_index = address.start + block_offset
      local block = blocks[block_index]
      local next = block_index < address.stop and (block_index + 1) or 0
      local prev = block_index > address.start and (block_index - 1) or 0
      write_u32_le(block, 4, next)
      write_u32_le(block, 8, prev)
      local from = block_offset * Format.BLOCK_ITEM_MAX_COUNT
      local to = math.min(#entries, from + Format.BLOCK_ITEM_MAX_COUNT)
      local position = Format.BLOCK_HEAD
      for entry_index = from + 1, to do
        local entry = entries[entry_index]
        block[position + 1] = entry.is_file and 1 or 0
        position = position + 1
        write_fixed_name(block, position, entry.name)
        position = position + Format.BLOCK_NAME_MAX_SIZE
        write_u32_le(block, position, entry.start)
        position = position + 4
        write_u32_le(block, position, entry.stop)
        position = position + 4
        write_u16_le(block, position, entry.offset)
        position = position + 2
      end
    end

    for _, child_entry in ipairs(dir.children) do
      local child = child_entry.node
      if child.kind == "dir" then
        serialize_dir(child)
      else
        serialize_file(child)
      end
    end
  end

  serialize_dir(root)
  return blocks
end

function Format.parse_tree(blocks)
  assert(#blocks >= 1 or blocks[1] ~= nil or blocks[0] ~= nil, "image must contain blocks")
  -- blocks is 0-indexed sparse array; length via highest index is awkward — use total from header
  local visited = {}

  local function read_block(index)
    local block = blocks[index]
    assert(block, "missing block " .. tostring(index))
    return block
  end

  local function parse_file(start, stop, offset)
    if start == 0 and stop == 0 and offset == 0 then
      return {}
    end
    assert(start > 0 and stop >= start, "invalid file block range")
    assert(offset >= Format.BLOCK_HEAD and offset <= Format.BLOCK_SIZE, "invalid file offset")
    local out = {}
    local current = start
    while true do
      local block = read_block(current)
      local limit = current == stop and offset or Format.BLOCK_SIZE
      assert(limit >= Format.BLOCK_HEAD and limit <= Format.BLOCK_SIZE, "invalid block limit")
      for i = Format.BLOCK_HEAD, limit - 1 do
        out[#out + 1] = block[i + 1]
      end
      if current == stop then
        break
      end
      current = read_u32_le(block, 4)
      assert(current ~= 0, "unterminated file block chain")
    end
    return out
  end

  local function parse_dir(index, name, parent)
    assert(not visited[index], "directory cycle at block " .. tostring(index))
    visited[index] = true
    local first_block = read_block(index)
    local dot = read_entry(first_block, Format.BLOCK_HEAD)
    assert(not dot.is_file and dot.name == ".", "invalid directory header at block " .. tostring(index))
    local dir = Format.new_dir(name, parent)
    local entries = {}
    local current = index
    while true do
      local block = read_block(current)
      local limit = current == dot.stop and dot.offset or Format.BLOCK_SIZE
      assert(limit >= Format.BLOCK_HEAD and limit <= Format.BLOCK_SIZE, "invalid directory offset")
      local position = Format.BLOCK_HEAD
      while position + Format.ENTRY_SIZE <= limit do
        entries[#entries + 1] = read_entry(block, position)
        position = position + Format.ENTRY_SIZE
      end
      if current == dot.stop then
        break
      end
      current = read_u32_le(block, 4)
      assert(current ~= 0, "unterminated directory chain")
    end
    assert(#entries >= 2, "directory missing dot entries")
    assert(entries[2].name == "..", "directory missing parent entry")
    for i = 3, #entries do
      local entry = entries[i]
      assert(entry.name ~= "", "empty child entry")
      local child
      if entry.is_file then
        child = Format.new_file(entry.name, dir, parse_file(entry.start, entry.stop, entry.offset))
      else
        child = parse_dir(entry.start, entry.name, dir)
      end
      Format.dir_set_child(dir, entry.name, child)
    end
    return dir
  end

  return parse_dir(Format.ROOT_BLOCKINDEX, "", nil)
end

function Format.blocks_to_string(blocks)
  -- blocks is 0-indexed; find max index
  local max_index = -1
  for i, _ in pairs(blocks) do
    if type(i) == "number" and i > max_index then
      max_index = i
    end
  end
  local parts = {}
  for i = 0, max_index do
    local block = blocks[i]
    local chars = {}
    for j = 1, Format.BLOCK_SIZE do
      chars[j] = string.char(block[j])
    end
    parts[#parts + 1] = table.concat(chars)
  end
  return table.concat(parts)
end

function Format.string_to_blocks(data)
  assert(#data % Format.BLOCK_SIZE == 0, "image size is not a multiple of 512")
  local total = #data // Format.BLOCK_SIZE
  local blocks = {}
  for i = 0, total - 1 do
    local block = empty_block()
    local base = i * Format.BLOCK_SIZE
    for j = 1, Format.BLOCK_SIZE do
      block[j] = string.byte(data, base + j)
    end
    blocks[i] = block
  end
  return blocks
end

function Format.read_u32_le_str(data, offset0)
  local i = offset0 + 1
  local b1, b2, b3, b4 = string.byte(data, i, i + 3)
  return b1 | (b2 << 8) | (b3 << 16) | (b4 << 24)
end

Format.write_u32_le = write_u32_le
Format.read_u32_le = read_u32_le
Format.write_u16_le = write_u16_le
Format.read_u16_le = read_u16_le
Format.read_fixed_name = read_fixed_name
Format.write_fixed_name = write_fixed_name
Format.empty_block = empty_block
Format.copy_bytes = copy_bytes

return Format

File: kotlin/src/main/kotlin/filefs/internal/Format.kt
package filefs.internal

import filefs.DirEntry
import filefs.FileType
import java.nio.charset.StandardCharsets
import java.util.IdentityHashMap
import kotlin.math.ceil

internal object FormatConstants {
    const val BLOCK_SIZE = 512
    const val BLOCK_ITEM_MAX_COUNT = 20
    const val BLOCK_HEAD = 12
    const val BLOCK_NAME_MAX_SIZE = 14
    const val ENTRY_SIZE = 25
    const val ROOT_BLOCKINDEX = 1
    const val DATA_PER_BLOCK = BLOCK_SIZE - BLOCK_HEAD

    val MAGIC: ByteArray = byteArrayOf(0x78, 0x11, 0x45, 0x14)
}

internal sealed class FsNode(open var name: String) {
    abstract var parent: DirNode?
}

internal class DirNode(
    override var name: String,
    override var parent: DirNode?,
) : FsNode(name) {
    val children: LinkedHashMap<String, FsNode> = LinkedHashMap()
}

internal class FileNode(
    override var name: String,
    override var parent: DirNode?,
    var data: ByteArray = ByteArray(0),
) : FsNode(name)

internal data class TreeCopy(
    val root: DirNode,
    val cwd: DirNode,
)

internal fun deepCopyTree(root: DirNode, cwd: DirNode): TreeCopy {
    val dirMap = IdentityHashMap<DirNode, DirNode>()

    fun copyDir(source: DirNode, parent: DirNode?): DirNode {
        val copy = DirNode(source.name, parent)
        dirMap[source] = copy
        for ((name, child) in source.children) {
            copy.children[name] = when (child) {
                is DirNode -> copyDir(child, copy)
                is FileNode -> FileNode(child.name, copy, child.data.copyOf())
            }
        }
        return copy
    }

    val rootCopy = copyDir(root, null)
    val cwdCopy = dirMap[cwd] ?: rootCopy
    return TreeCopy(rootCopy, cwdCopy)
}

internal fun buildAbsolutePath(dir: DirNode): String {
    if (dir.parent == null) {
        return "/"
    }

    val parts = ArrayDeque<String>()
    var current: DirNode? = dir
    while (current != null && current.parent != null) {
        parts.addFirst(current.name)
        current = current.parent
    }

    return buildString {
        append('/')
        while (parts.isNotEmpty()) {
            append(parts.removeFirst())
            append('/')
        }
    }
}

internal fun listDirectoryEntries(dir: DirNode): List<DirEntry> {
    val entries = ArrayList<DirEntry>(dir.children.size + 2)
    val dotType = if (dir.parent == null) FileType.ROOT else FileType.DIR
    val dotDotType = if (dir.parent == null) FileType.ROOT else FileType.DIR
    entries += DirEntry(dotType, ".")
    entries += DirEntry(dotDotType, "..")
    for (child in dir.children.values) {
        val type = when (child) {
            is DirNode -> FileType.DIR
            is FileNode -> FileType.FILE
        }
        entries += DirEntry(type, child.name)
    }
    return entries
}

internal data class SerializedImage(
    val blocks: List<ByteArray>,
)

internal fun serializeTree(root: DirNode): SerializedImage {
    val addresses = IdentityHashMap<FsNode, NodeAddress>()

    fun assign(node: FsNode, nextIndex: Int): Int {
        return when (node) {
            is DirNode -> {
                val entryCount = 2 + node.children.size
                val blockCount = ceil(entryCount / FormatConstants.BLOCK_ITEM_MAX_COUNT.toDouble()).toInt()
                val start = nextIndex
                val stop = start + blockCount - 1
                val remainder = entryCount % FormatConstants.BLOCK_ITEM_MAX_COUNT
                val offset = if (remainder == 0) {
                    FormatConstants.BLOCK_SIZE
                } else {
                    FormatConstants.BLOCK_HEAD + remainder * FormatConstants.ENTRY_SIZE
                }
                addresses[node] = NodeAddress(start, stop, offset, blockCount)
                var cursor = stop + 1
                for (child in node.children.values) {
                    cursor = assign(child, cursor)
                }
                cursor
            }

            is FileNode -> {
                if (node.data.isEmpty()) {
                    addresses[node] = NodeAddress(0, 0, 0, 0)
                    nextIndex
                } else {
                    val blockCount = ceil(node.data.size / FormatConstants.DATA_PER_BLOCK.toDouble()).toInt()
                    val start = nextIndex
                    val stop = start + blockCount - 1
                    val remainder = node.data.size % FormatConstants.DATA_PER_BLOCK
                    val offset = if (remainder == 0) {
                        FormatConstants.BLOCK_SIZE
                    } else {
                        FormatConstants.BLOCK_HEAD + remainder
                    }
                    addresses[node] = NodeAddress(start, stop, offset, blockCount)
                    stop + 1
                }
            }
        }
    }

    val totalBlocks = assign(root, FormatConstants.ROOT_BLOCKINDEX)
    val blocks = MutableList(totalBlocks) { ByteArray(FormatConstants.BLOCK_SIZE) }

    writeBytes(blocks[0], 0, FormatConstants.MAGIC)
    writeU32LE(blocks[0], 4, totalBlocks)
    writeU32LE(blocks[0], 8, 0)

    fun serializeFile(file: FileNode) {
        val address = addresses[file] ?: error("missing address for file")
        if (address.start == 0) {
            return
        }

        var sourceOffset = 0
        for (blockIndex in address.start..address.stop) {
            val block = blocks[blockIndex]
            val next = if (blockIndex < address.stop) blockIndex + 1 else 0
            val prev = if (blockIndex > address.start) blockIndex - 1 else 0
            writeU32LE(block, 4, next)
            writeU32LE(block, 8, prev)

            val endExclusive = minOf(sourceOffset + FormatConstants.DATA_PER_BLOCK, file.data.size)
            file.data.copyInto(
                destination = block,
                destinationOffset = FormatConstants.BLOCK_HEAD,
                startIndex = sourceOffset,
                endIndex = endExclusive,
            )
            sourceOffset = endExclusive
        }
    }

    fun serializeDir(dir: DirNode) {
        val address = addresses[dir] ?: error("missing address for directory")
        val entries = ArrayList<WireEntry>(dir.children.size + 2)
        entries += WireEntry(
            isFile = false,
            name = ".",
            start = address.start,
            stop = address.stop,
            offset = address.offset,
        )
        entries += WireEntry(
            isFile = false,
            name = "..",
            start = dir.parent?.let { addresses[it]?.start } ?: 0,
            stop = 0,
            offset = 0,
        )
        for (child in dir.children.values) {
            val childAddress = addresses[child] ?: error("missing address for child")
            entries += WireEntry(
                isFile = child is FileNode,
                name = child.name,
                start = childAddress.start,
                stop = childAddress.stop,
                offset = childAddress.offset,
            )
        }

        for (blockOffset in 0 until address.blockCount) {
            val blockIndex = address.start + blockOffset
            val block = blocks[blockIndex]
            val next = if (blockIndex < address.stop) blockIndex + 1 else 0
            val prev = if (blockIndex > address.start) blockIndex - 1 else 0
            writeU32LE(block, 4, next)
            writeU32LE(block, 8, prev)

            val from = blockOffset * FormatConstants.BLOCK_ITEM_MAX_COUNT
            val to = minOf(entries.size, from + FormatConstants.BLOCK_ITEM_MAX_COUNT)
            var position = FormatConstants.BLOCK_HEAD
            for (entry in entries.subList(from, to)) {
                block[position] = if (entry.isFile) 1 else 0
                position += 1
                writeFixedName(block, position, entry.name)
                position += FormatConstants.BLOCK_NAME_MAX_SIZE
                writeU32LE(block, position, entry.start)
                position += 4
                writeU32LE(block, position, entry.stop)
                position += 4
                writeU16LE(block, position, entry.offset)
                position += 2
            }
        }

        for (child in dir.children.values) {
            when (child) {
                is DirNode -> serializeDir(child)
                is FileNode -> serializeFile(child)
            }
        }
    }

    serializeDir(root)
    return SerializedImage(blocks)
}

internal fun parseTree(blocks: List<ByteArray>): DirNode {
    require(blocks.size >= 2) { "image must contain at least two blocks" }

    val visited = mutableSetOf<Int>()

    fun readBlock(index: Int): ByteArray {
        return blocks.getOrNull(index) ?: throw IllegalArgumentException("missing block $index")
    }

    fun parseFile(start: Int, stop: Int, offset: Int): ByteArray {
        if (start == 0 && stop == 0 && offset == 0) {
            return ByteArray(0)
        }
        require(start > 0 && stop >= start) { "invalid file block range" }
        require(offset in FormatConstants.BLOCK_HEAD..FormatConstants.BLOCK_SIZE) { "invalid file offset" }

        val out = ArrayList<Byte>()
        var current = start
        while (true) {
            val block = readBlock(current)
            val limit = if (current == stop) offset else FormatConstants.BLOCK_SIZE
            require(limit in FormatConstants.BLOCK_HEAD..FormatConstants.BLOCK_SIZE) { "invalid block limit" }
            for (i in FormatConstants.BLOCK_HEAD until limit) {
                out += block[i]
            }
            if (current == stop) {
                break
            }
            current = readU32LE(block, 4)
            require(current != 0) { "unterminated file block chain" }
        }
        return out.toByteArray()
    }

    fun parseDir(index: Int, name: String, parent: DirNode?): DirNode {
        require(visited.add(index)) { "directory cycle at block $index" }
        val firstBlock = readBlock(index)
        val dot = readEntry(firstBlock, FormatConstants.BLOCK_HEAD)
        require(!dot.isFile && dot.name == ".") { "invalid directory header at block $index" }

        val dir = DirNode(name, parent)
        val entries = ArrayList<WireEntry>()
        var current = index
        while (true) {
            val block = readBlock(current)
            val limit = if (current == dot.stop) dot.offset else FormatConstants.BLOCK_SIZE
            require(limit in FormatConstants.BLOCK_HEAD..FormatConstants.BLOCK_SIZE) { "invalid directory offset" }
            var position = FormatConstants.BLOCK_HEAD
            while (position + FormatConstants.ENTRY_SIZE <= limit) {
                entries += readEntry(block, position)
                position += FormatConstants.ENTRY_SIZE
            }
            if (current == dot.stop) {
                break
            }
            current = readU32LE(block, 4)
            require(current != 0) { "unterminated directory chain" }
        }

        require(entries.size >= 2) { "directory missing dot entries" }
        require(entries[1].name == "..") { "directory missing parent entry" }

        for (entry in entries.drop(2)) {
            require(entry.name.isNotEmpty()) { "empty child entry" }
            val child = if (entry.isFile) {
                FileNode(entry.name, dir, parseFile(entry.start, entry.stop, entry.offset))
            } else {
                parseDir(entry.start, entry.name, dir)
            }
            dir.children[child.name] = child
        }
        return dir
    }

    return parseDir(FormatConstants.ROOT_BLOCKINDEX, "", null)
}

internal fun readU32LE(bytes: ByteArray, offset: Int): Int {
    return (bytes[offset].toInt() and 0xFF) or
        ((bytes[offset + 1].toInt() and 0xFF) shl 8) or
        ((bytes[offset + 2].toInt() and 0xFF) shl 16) or
        ((bytes[offset + 3].toInt() and 0xFF) shl 24)
}

internal fun writeU32LE(bytes: ByteArray, offset: Int, value: Int) {
    bytes[offset] = (value and 0xFF).toByte()
    bytes[offset + 1] = ((value ushr 8) and 0xFF).toByte()
    bytes[offset + 2] = ((value ushr 16) and 0xFF).toByte()
    bytes[offset + 3] = ((value ushr 24) and 0xFF).toByte()
}

internal fun readU16LE(bytes: ByteArray, offset: Int): Int {
    return (bytes[offset].toInt() and 0xFF) or ((bytes[offset + 1].toInt() and 0xFF) shl 8)
}

internal fun writeU16LE(bytes: ByteArray, offset: Int, value: Int) {
    bytes[offset] = (value and 0xFF).toByte()
    bytes[offset + 1] = ((value ushr 8) and 0xFF).toByte()
}

internal fun writeBytes(destination: ByteArray, offset: Int, value: ByteArray) {
    value.copyInto(destination, offset)
}

internal fun readFixedName(bytes: ByteArray, offset: Int): String {
    var end = offset
    val stop = offset + FormatConstants.BLOCK_NAME_MAX_SIZE
    while (end < stop && bytes[end] != 0.toByte()) {
        end += 1
    }
    return bytes.copyOfRange(offset, end).toString(StandardCharsets.US_ASCII)
}

internal fun writeFixedName(bytes: ByteArray, offset: Int, name: String) {
    require(name.length <= FormatConstants.BLOCK_NAME_MAX_SIZE) { "name exceeds 14 bytes: $name" }
    repeat(FormatConstants.BLOCK_NAME_MAX_SIZE) { index ->
        bytes[offset + index] = 0
    }
    name.toByteArray(StandardCharsets.US_ASCII).copyInto(bytes, offset)
}

private fun readEntry(bytes: ByteArray, offset: Int): WireEntry {
    val isFile = (bytes[offset].toInt() and 0x01) == 1
    val name = readFixedName(bytes, offset + 1)
    val start = readU32LE(bytes, offset + 1 + FormatConstants.BLOCK_NAME_MAX_SIZE)
    val stop = readU32LE(bytes, offset + 1 + FormatConstants.BLOCK_NAME_MAX_SIZE + 4)
    val entryOffset = readU16LE(bytes, offset + 1 + FormatConstants.BLOCK_NAME_MAX_SIZE + 8)
    return WireEntry(isFile, name, start, stop, entryOffset)
}

private data class NodeAddress(
    val start: Int,
    val stop: Int,
    val offset: Int,
    val blockCount: Int,
)

private data class WireEntry(
    val isFile: Boolean,
    val name: String,
    val start: Int,
    val stop: Int,
    val offset: Int,
)
